src/plot_api/plot_api.js

```diff
@@ -153,6 +153,9 @@ Plotly.plot = function(gd, data, layout, config) {
         makePlotFramework(gd);
     }
 
+    // clear gradient defs on each .plot call, because we know we'll loop through all traces
+    Drawing.initGradients(gd);
```

Comment:
It would be nice to add jasmine test checking that `restyle` can update gradient `defs` properly.